### PR TITLE
Added configuration option to set Conviva Device Metadata

### PR DESCRIPTION
**Pull request by cristian-atehortua**
Solves #60.

Same as #61 but created from version 4.0.2

### CODE REVIEW

File: package.json
Comment (dweinber) [inline package.json:42]: I think we don't need the `bitmovin-player` in both, `devDependencies` and `peerDependencies`, right? I think having it in `peerDependencies` makes sense, so I'd suggestion removing it from `devDependencies` if there are no good reasons to keep it there.

File: src/ts/Conviva.d.ts
Comment (dweinber) [inline src/ts/Conviva.d.ts:149]: Why do we redefine a rather weak type here (Why do we need the `undefined`, btw - should be avoided in types unless really necessary)? We have this properly defined in an interface in `ConvivaAnalytics.ts`. If we'd use that, the user would have a better type checking and potentially auto-completion to create the metadata object.

File: src/ts/ConvivaAnalytics.ts
Comment (dweinber) [inline src/ts/ConvivaAnalytics.ts:170]: Is it recommended by Conviva to always send all of the `deviceMetadata` fields, even if they might be `undefined`?

**Review by dweinber — CHANGES_REQUESTED** [review]

File: package.json
Comment (cristian-atehortua) [reply package.json:42]: `peerDependencies` are not installed when you run `npm install`. They are useful to tell the consumers what dependencies your package need to be able to run. It is needed to keep `bitmovin-player` in `devDependencies` to be able to import the types from that package.

File: src/ts/ConvivaAnalytics.ts
Comment (cristian-atehortua) [reply src/ts/ConvivaAnalytics.ts:170]: In practice, for Conviva library, sending them as undefined is the same that not sending them at all. We need to define in this way because the user can provide those values or not.

File: src/ts/Conviva.d.ts
Comment (cristian-atehortua) [reply src/ts/Conviva.d.ts:149]: If you see where this type is used and the comparison with previous changes you will see that they are pretty similar. The difference is supporting other types for the values of this object because with new device metadata options can have other types.
Regarding the use of `undefined` it could be deleted.

**Comment by cristian-atehortua** [comment]
@dweinber Hi, I answered your comments and updated the PR with a change related to them, can you review it again?

File: src/ts/ConvivaAnalytics.ts
Comment (stonko1994) [inline src/ts/ConvivaAnalytics.ts:35]: I think we should deprecate this property here in favor our the `deviceMetadata.category`

File: CHANGELOG.md
Comment (stonko1994) [inline CHANGELOG.md]:
```suggestion
- Possibility to set detailed Conviva Device Metadata via `ConvivaAnalyticsConfiguration.deviceMetadata`
```

If we going to deprecate the existing one please add a `### Deprecated` section.

**Review by stonko1994 — COMMENTED** [review]
@dweinber what do you think about deprecating the existing property? 🤔

File: src/ts/ConvivaAnalytics.ts
Comment (cristian-atehortua) [reply src/ts/ConvivaAnalytics.ts:35]: Hi, @stonko1994. Please confirm if should I deprecate that field and what is the process you have to do it. (Just adding a `@deprecated` JSDoc tag? Adding a warning to the console when this field is used?)

**Comment by cristian-atehortua** [comment]
@stonko1994 @dweinber Can you please check out the comments and answer them? Thanks

**Review by stonko1994 — COMMENTED** [review]
From my point of view, we should deprecate it however we need the input from @dweinber as well.

File: src/ts/ConvivaAnalytics.ts
Comment (stonko1994) [reply src/ts/ConvivaAnalytics.ts:35]: Yes, just a `@deprecated` JSDoc tag is enough 🙁

File: package.json
Comment (dweinber) [reply package.json:42]: Ah right 👍  Thanks

Shouldn't we use the same version actually in the `devDependencies` (i.e. adding `^`)?

File: src/ts/Conviva.d.ts
Comment (dweinber) [reply src/ts/Conviva.d.ts:149]: Having spent a bit of time on this now I agree, let's keep it as it is 😅

File: src/ts/ConvivaAnalytics.ts
Comment (dweinber) [reply src/ts/ConvivaAnalytics.ts:170]: I see, thanks for explaining. Let's keep it this way, keeps the implementation simple 👍

**Review by dweinber — APPROVED** [review]
Looks good now 👍  
Please just have a look at my comment in `package.json`.

@stonko1994 good idea to deprecate the old field, thanks.

File: package.json
Comment (cristian-atehortua) [reply package.json:42]: Yes, I will update it.

**Comment by cristian-atehortua** [comment]
@dweinber I just updated with your last comment.

Thank you both @stonko1994 and @dweinber  for your help. If you don't have any other comments you can merge when you want.

**Comment by dweinber** [comment]
Great, thanks!
I'll just wait for @stonko1994's approval (or change request) and then we're good to merge.

**Review by stonko1994 — APPROVED** [review]
LGTM 👍 Feel free to merge 🙂

**Comment by cristian-atehortua** [comment]
@dweinber @stonko1994 Thanks again.
I think one of you should merge cause "Only those with write access to this repository can merge pull requests."

image

**Pull request merged**
